controllers/controllers/networking/security_groups/controller.go

```diff
@@ -0,0 +1,473 @@
+		}
+
+		egressRules = append(egressRules, v3.Rule{
+			Action:   v3.Allow,
+			Protocol: &numorstring.Protocol{Type: 1, StrVal: getProtocol(rule.Protocol)},
+			Destination: v3.EntityRule{
+				Nets:  nets,
+				Ports: ports,
+			},
+		})
+	}
+
+	return egressRules, nil
+}
+
+func buildRulePorts(rule korifiv1alpha1.SecurityGroupRule) ([]numorstring.Port, error) {
+	var ports []numorstring.Port
+
+	if strings.Contains(rule.Ports, "-") {
+		port, err := parseRangePorts(rule.Ports)
+		if err != nil {
+			return nil, err
+		}
+
+		return []numorstring.Port{port}, nil
+	}
+
+	for _, portStr := range strings.Split(rule.Ports, ",") {
+		port, err := portStringToUint16(portStr)
+		if err != nil {
+			return nil, err
+		}
+
+		ports = append(ports, numorstring.Port{MinPort: port, MaxPort: port})
+	}
+
+	return ports, nil
+}
+
+func buildSelector(workloads korifiv1alpha1.SecurityGroupWorkloads) string {
+	var workloadTypes []string
+
+	if workloads.Running {
+		workloadTypes = append(workloadTypes, korifiv1alpha1.CFWorkloadTypeApp)
+	}
+	if workloads.Staging {
+		workloadTypes = append(workloadTypes, korifiv1alpha1.CFWorkloadTypeBuild)
+	}
+	values := "'" + strings.Join(workloadTypes, "', '") + "'"
+
+	return fmt.Sprintf("%s in { %s }", korifiv1alpha1.CFWorkloadTypeLabelkey, values)
+}
+
+func parseRangePorts(ports string) (numorstring.Port, error) {
+	rangePorts := strings.Split(ports, "-")
+	if len(rangePorts) != 2 {
+		return numorstring.Port{}, fmt.Errorf("invalid port range format: %s", ports)
+	}
+
+	start, err := portStringToUint16(rangePorts[0])
+	if err != nil {
+		return numorstring.Port{}, err
+	}
+
+	end, err := portStringToUint16(rangePorts[1])
+	if err != nil {
+		return numorstring.Port{}, err
+	}
+
+	return numorstring.Port{
+		MinPort: start,
+		MaxPort: end,
+	}, nil
+}
+
+func portStringToUint16(port string) (uint16, error) {
+	pStr := strings.TrimSpace(port)
+	if pStr == "" {
+		return 0, fmt.Errorf("port value cannot be empty")
+	}
+	p, err := strconv.ParseUint(pStr, 10, 16)
+	if err != nil {
+		return 0, fmt.Errorf("invalid port %s: %w", pStr, err)
+	}
+	return uint16(p), nil
+}
+
+func buildRuleNets(rule korifiv1alpha1.SecurityGroupRule) ([]string, error) {
+	if strings.Contains(rule.Destination, "-") {
+		var nets []string
+		parts := strings.Split(rule.Destination, "-")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid IP range format: %s", rule.Destination)
+		}
+
+		cidrs, err := generateCIDRs(parts[0], parts[1])
+		if err != nil {
+			return nil, err
+		}
+		nets = append(nets, cidrs...)
+
+		return nets, nil
+	}
+
+	if _, _, err := net.ParseCIDR(rule.Destination); err == nil {
+		return []string{rule.Destination}, nil
+	}
+
+	if ip := net.ParseIP(rule.Destination); ip != nil {
+		return []string{fmt.Sprintf("%s/32", rule.Destination)}, nil
+	}
+
+	return nil, fmt.Errorf("invalid destination: %s", rule.Destination)
+}
+
+func generateCIDRs(startIP, endIP string) ([]string, error) {
+	start, err := ipToUint32(startIP)
```

Comment:
Would it make sense to move parsing and validating various ports/ips,etc into a validating webhook thus simplifying the controller code.